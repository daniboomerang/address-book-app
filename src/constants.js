const SWISS = 'CH';
const SPANISH = 'ES';
const BRITISH = 'BR';
const FRENCH = 'FR';

export { SWISS, SPANISH, BRITISH, FRENCH };
