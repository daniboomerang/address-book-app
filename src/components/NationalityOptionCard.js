import React from 'react';
import cx from 'classnames';
import { SWISS, SPANISH, BRITISH } from '../constants';
import { BrFlag, ChFlag, EsFlag, FrFlag } from './flags';

/**
 * @param {Props} props
 * @param {string} props.nationality - The different nationality options -> CH, ES, BR or FR
 * @param {boolean} props.isSelected - Whether the options shall be deisplayed as selected
 * @param {Function} props.addNationality - Action to add a new nationality to the store
 * @param {Function} props.removeNationality - Action to remove a nationality from the store
 */
const NationalityOptionCard = ({ nationality, isSelected, addNationality, removeNationality }) => {
  const handleAddNationality = () => addNationality(nationality);
  const handleRemoveNationality = () => removeNationality(nationality);

  const renderFlag = () => {
    if (nationality === BRITISH) {
      return <BrFlag className="h-12 w-12 sm:h-24 sm:w-24" />;
    }
    if (nationality === SWISS) {
      return <ChFlag className="h-12 w-12 sm:h-24 sm:w-24" />;
    }
    if (nationality === SPANISH) {
      return <EsFlag className="h-12 w-12 sm:h-24 sm:w-24" />;
    }
    return <FrFlag className="h-12 w-12 sm:h-24 sm:w-24" />;
  };

  return (
    <button
      type="button"
      data-testid={`nationality-${nationality}`}
      onClick={isSelected ? handleRemoveNationality : handleAddNationality}
      className={cx(
        'focus:outline-none grid border-4 text-center justify-center transform cursor-pointer duration-200 hover:border-red-600 hover:text-red-600 pt-4 pb-2 md:pt-5 md:pb-3 w-24 sm:w-32',
        {
          'border-red-600 text-red-600 rounded-md scale-110 hover:border-blue-500 hover:text-blue-500': isSelected,
          'hover:rounded-md border-blue-500': !isSelected,
        }
      )}
    >
      {renderFlag()}
      <span className="mt-4">{nationality}</span>
    </button>
  );
};

export default NationalityOptionCard;
