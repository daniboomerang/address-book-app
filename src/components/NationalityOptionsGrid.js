import React, { useState } from 'react';
import { SWISS, SPANISH, BRITISH, FRENCH } from '../constants';
import NationalityOptionCard from './NationalityOptionCard';

/**
 * It renders a grid with the different nationality option cards
 */
const NationalityOptionsGrid = () => {
  const [selectedNationalities] = useState([SWISS, SPANISH, FRENCH]);

  return (
    <div className="p-4 max-w-md m-auto">
      <div className="my-12 sm:my-20 text-center">
        Select the nationalities you would like to filter
        <br />
        <div className="text-sm">{`(${selectedNationalities.toString()})`}</div>
      </div>

      <div
        className="grid grid-cols-2 justify-between items-center m-auto col-gap-4 row-gap-12"
        style={{ justifyItems: 'center' }}
      >
        <NationalityOptionCard
          nationality={SWISS}
          isSelected={selectedNationalities.includes(SWISS)}
        />
        <NationalityOptionCard
          nationality={SPANISH}
          isSelected={selectedNationalities.includes(SPANISH)}
        />
        <NationalityOptionCard
          nationality={BRITISH}
          isSelected={selectedNationalities.includes(BRITISH)}
        />
        <NationalityOptionCard
          nationality={FRENCH}
          isSelected={selectedNationalities.includes(FRENCH)}
        />
      </div>
    </div>
  );
};

export default NationalityOptionsGrid;
