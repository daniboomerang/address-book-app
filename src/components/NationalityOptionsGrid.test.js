import React from 'react';
import { mount } from 'enzyme';
import NationalityOptionsGrid from './NationalityOptionsGrid';
import { SWISS, SPANISH, BRITISH, FRENCH } from '../constants';
import NationalityOptionCard from './NationalityOptionCard';

describe('NationalityOptionsGrid component', () => {
  it('should correctly render all nationality option cards', () => {
    const wrapper = mount(<NationalityOptionsGrid />);

    expect(wrapper.find('NationalityOptionCard').length).toBe(4);
    expect(wrapper).toMatchSnapshot();
  });
});
