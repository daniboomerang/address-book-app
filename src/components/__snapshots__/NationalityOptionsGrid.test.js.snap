// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NationalityOptionsGrid component should correctly render all nationality option cards 1`] = `
<NationalityOptionsGrid>
  <div
    className="p-4 max-w-md m-auto"
  >
    <div
      className="my-12 sm:my-20 text-center"
    >
      Select the nationalities you would like to filter
      <br />
      <div
        className="text-sm"
      >
        (CH,ES,FR)
      </div>
    </div>
    <div
      className="grid grid-cols-2 justify-between items-center m-auto col-gap-4 row-gap-12"
      style={
        Object {
          "justifyItems": "center",
        }
      }
    >
      <NationalityOptionCard
        isSelected={true}
        nationality="CH"
      >
        <div
          className="grid border-4 text-center justify-center transform cursor-pointer duration-200 hover:border-red-600 hover:text-red-600 pt-4 pb-2 md:pt-5 md:pb-3 w-24 md:w-32 border-red-600 text-red-600 rounded-md scale-110 hover:scale-100 hover:border-blue-500"
          data-testid="nationality-CH"
        >
          <svg
            className="h-12 w-12 md:h-24 md:w-24"
            version="1.1"
            viewBox="0 0 512 512"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#D80027"
              height="512"
              width="512"
            />
            <polygon
              fill="#F0F0F0"
              points="395.13,209.624 302.377,209.624 302.377,116.87 209.623,116.87 209.623,209.624   116.87,209.624 116.87,302.377 209.623,302.377 209.623,395.13 302.377,395.13 302.377,302.377 395.13,302.377 "
            />
          </svg>
          <span
            className="mt-4"
          >
            CH
          </span>
        </div>
      </NationalityOptionCard>
      <NationalityOptionCard
        isSelected={true}
        nationality="ES"
      >
        <div
          className="grid border-4 text-center justify-center transform cursor-pointer duration-200 hover:border-red-600 hover:text-red-600 pt-4 pb-2 md:pt-5 md:pb-3 w-24 md:w-32 border-red-600 text-red-600 rounded-md scale-110 hover:scale-100 hover:border-blue-500"
          data-testid="nationality-ES"
        >
          <svg
            className="h-12 w-12 md:h-24 md:w-24"
            version="1.1"
            viewBox="0 0 512 512"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#FFDA44"
              height="512"
              width="512"
            />
            <g>
              <rect
                fill="#D80027"
                height="170.663"
                width="512"
              />
              <rect
                fill="#D80027"
                height="170.663"
                width="512"
                y="341.337"
              />
            </g>
          </svg>
          <span
            className="mt-4"
          >
            ES
          </span>
        </div>
      </NationalityOptionCard>
      <NationalityOptionCard
        isSelected={false}
        nationality="BR"
      >
        <div
          className="grid border-4 text-center justify-center transform cursor-pointer duration-200 hover:border-red-600 hover:text-red-600 pt-4 pb-2 md:pt-5 md:pb-3 w-24 md:w-32 hover:scale-110 hover:rounded-md border-blue-500"
          data-testid="nationality-BR"
        >
          <svg
            className="h-12 w-12 md:h-24 md:w-24"
            version="1.1"
            viewBox="0 0 512 512"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#F0F0F0"
              height="512"
              width="512"
            />
            <polygon
              fill="#D80027"
              points="0,304 208,304 208,512 304,512 304,304 512,304 512,208 304,208 304,0 208,0 208,208 0,208   "
            />
            <g>
              <polygon
                fill="#0052B4"
                points="406.92,333.913 512,438.993 512,333.913  "
              />
              <polygon
                fill="#0052B4"
                points="333.913,333.913 512,512 512,461.64 384.273,333.913  "
              />
              <polygon
                fill="#0052B4"
                points="464.564,512 333.913,381.336 333.913,512  "
              />
            </g>
            <polygon
              fill="#F0F0F0"
              points="333.913,333.913 512,512 512,461.64 384.273,333.913 "
            />
            <polygon
              fill="#D80027"
              points="333.913,333.913 512,512 512,461.64 384.273,333.913 "
            />
            <g>
              <polygon
                fill="#0052B4"
                points="80.302,333.913 0,414.215 0,333.913  "
              />
              <polygon
                fill="#0052B4"
                points="178.084,356.559 178.084,511.997 22.658,511.997  "
              />
            </g>
            <polygon
              fill="#D80027"
              points="127.724,333.916 0,461.641 0,512 0,512 178.084,333.916 "
            />
            <g>
              <polygon
                fill="#0052B4"
                points="105.08,178.087 0,73.007 0,178.087  "
              />
              <polygon
                fill="#0052B4"
                points="178.087,178.087 0,0 0,50.36 127.727,178.087  "
              />
              <polygon
                fill="#0052B4"
                points="47.436,0 178.087,130.663 178.087,0  "
              />
            </g>
            <polygon
              fill="#F0F0F0"
              points="178.087,178.087 0,0 0,50.36 127.727,178.087 "
            />
            <polygon
              fill="#D80027"
              points="178.087,178.087 0,0 0,50.36 127.727,178.087 "
            />
            <g>
              <polygon
                fill="#0052B4"
                points="431.698,178.087 512,97.785 512,178.087  "
              />
              <polygon
                fill="#0052B4"
                points="333.916,155.441 333.916,0.003 489.342,0.003  "
              />
            </g>
            <polygon
              fill="#D80027"
              points="384.276,178.084 512,50.359 512,0 512,0 333.916,178.084 "
            />
          </svg>
          <span
            className="mt-4"
          >
            BR
          </span>
        </div>
      </NationalityOptionCard>
      <NationalityOptionCard
        isSelected={true}
        nationality="FR"
      >
        <div
          className="grid border-4 text-center justify-center transform cursor-pointer duration-200 hover:border-red-600 hover:text-red-600 pt-4 pb-2 md:pt-5 md:pb-3 w-24 md:w-32 border-red-600 text-red-600 rounded-md scale-110 hover:scale-100 hover:border-blue-500"
          data-testid="nationality-FR"
        >
          <svg
            className="h-12 w-12 md:h-24 md:w-24"
            version="1.1"
            viewBox="0 0 512 512"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#F0F0F0"
              height="512"
              width="512"
            />
            <rect
              fill="#0052B4"
              height="512"
              width="170.663"
            />
            <rect
              fill="#D80027"
              height="512"
              width="170.663"
              x="341.337"
            />
          </svg>
          <span
            className="mt-4"
          >
            FR
          </span>
        </div>
      </NationalityOptionCard>
    </div>
  </div>
</NationalityOptionsGrid>
`;
